docs/proposals/resource-deletion-with-approval.md

```diff
@@ -0,0 +1,157 @@
+---
+title: Neat-enhancement-idea
+authors:
+- "@alexmt"
+  sponsors:
+- TBD
+  reviewers:
+- "@jessesuen"
+- TBD
+  approvers:
+- "@jessesuen"
+- TBD
+
+creation-date: 2020-04-19
+last-updated: 2020-04-19
+
+---
+# Neat Enhancement Idea
+
+Support manual approval for pruning and deleting Kubernetes resources during application syncing/deletion.
+
+## Summary
+
+Introduce Kubernetes resource-level annotations that require manual user approval using Argo CD UI/CLI/API before the
+resource is pruned or deleted. The annotations should be respected while Argo CD attempts to synchronize or delete the
+application.
+
+## Motivation
+
+We’ve seen cases where Argo CD deleted Kubernetes resources due to a bug or misconfiguration.​ Examples include [corrupted
+data](https://github.com/argoproj/argo-cd/issues/4423) in Redis, user errors
+([1](https://github.com/argoproj/argo-cd/issues/9093), [2](https://github.com/argoproj/argo-cd/issues/4844))
+and [bug](https://github.com/argoproj/argo-cd/issues/3473) in the automation on top of Argo CD. These examples don’t
+mean Argo CD is not reliable; however, there are cases where misbehavior is catastrophic, and erroneous deletion is not
+acceptable. Examples include the app-of-apps pattern where Argo CD is used to manage itself, or namespaces in production
+clusters.
+
+### Goals
+
+The goals of a proposal ares:
+
+#### Allow developers to mark resources that require manual approval before application deletion.
+
+Developers should be able to add an annotation to resources that require manual approval before deletion. The annotation
+should be respected by Argo CD when it attempts to delete the application.
+
+#### Allow developers to mark resources that require manual approval before pruning
+
+Developers should be able to add an annotation to resources that require manual approval before pruning. The annotation
+should be respected by Argo CD when it attempts to prune extra resources while syncing the application.
+
+### Non-Goals
+
+#### Implement automatic self check while deleting resources
+
+We've made our best effort to implement corrected behavior, and as of now, we are not aware of any bugs that cause
+erroneous deletion. The goal of this proposal is to provide a safety net for cases where deletion is not acceptable.
+
+## Proposal
```

Comment:
Will this feature provide protection for things besides sync operations? For example, will I be prompted if I attempt to manually delete a single resource via the UI/CLI? How about if I delete an App with cascading enabled?

Reply:
Yes, especially the last use case is interesting - how will this work without using the Argo CD API, e.g. in a fully declarative context?

Reply:
It would protect from both cases. We already have "sync" to control resource deletion during app deletion: `argocd.argoproj.io/sync-options: Delete=false` and another option for pruning during syncing: `argocd.argoproj.io/sync-options: Prune=false`. 

The proposal suggest to add `confirm` value for both. 

Added clarification to the proposal.

Reply:
Documented "declarative" approval as well.